bin/cheribsdtest/cheribsdtest_ipc.c

```diff
@@ -0,0 +1,85 @@
+/*-
+ * Copyright (c) 2021 Robert N. M. Watson
+ * All rights reserved.
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory (Department of Computer Science and
+ * Technology) under DARPA contract HR0011-18-C-0016 ("ECATS"), as part of the
+ * DARPA SSITH research programme.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <sys/types.h>
+#include <sys/sysctl.h>
+#include <sys/time.h>
+
+#include <cheri/cheri.h>
+#include <cheri/cheric.h>
+
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sysexits.h>
+#include <unistd.h>
+
+#include "cheribsdtest.h"
+
+/*
+ * As a regression test, block the current thread in a pipe write using a
+ * large write, and trigger a timer signal to interrupt that write.
+ *
+ * XXXRW: Update with more complete description of the bug once debugged and
+ * fixed.
+ */
+
+#define	BUFFER_SIZE	8192
+
+static void
+sigalrm_handler(int signum __unused)
+{
+
+}
+
+CHERIBSDTEST(test_ipc_pipe_sleep_signal,
+    "check that pipe IPC can be safely interrupted when sleeping on send")
+{
+	char buffer[BUFFER_SIZE];
+	ssize_t len;
+	int fds[2];
+
+	CHERIBSDTEST_CHECK_SYSCALL(signal(SIGALRM, sigalrm_handler));
+	CHERIBSDTEST_CHECK_SYSCALL(pipe(fds));
+	CHERIBSDTEST_CHECK_SYSCALL(alarm(1));
+	len = write(fds[0], buffer, sizeof(buffer));
+	if (len < 0 && errno != EINTR)
+		cheribsdtest_failure_err("write");
```

Comment:
Is it really a successful test if you write the entire buffer without getting interrupted?  If you want to ensure that the write was interrupted, then I think what you want to do instead might be something like:

```
static volatile sig_atomic_t alarm_fired;

static void
sigalrm_handler(...)
{
    alarm_fired = 1;
}

CHERIBSDTEST(...)
{
   ...
   len = write(fds[0], buffer, sizeof(buffer));
   if (len == -1 && errno != EINTR)
       cheribsdtest_failure_err("write");
   /* we probably have some sort of wrapper for this */
   if (alarm_fired == 0)
       cheribsdtest_failure("failed to interrupt write");
```

Even this is somewhat racy as in theory the handler could fire before or after the write(), but those races are probably rare enough to ignore.